xds/src/main/java/io/grpc/xds/GrpcAuthorizationEngine.java

```diff
@@ -0,0 +1,381 @@
+  abstract static class AuthDecision {
+    abstract Action decision();
+
+    @Nullable
+    abstract String matchingPolicyName();
+
+    static AuthDecision create(Action decisionType, @Nullable String matchingPolicy) {
+      return new AutoValue_GrpcAuthorizationEngine_AuthDecision(decisionType, matchingPolicy);
+    }
+  }
+
+  /** Represents authorization config policy that the engine will evaluate against. */
+  static final class AuthConfig {
+    private final List<PolicyMatcher> policies;
+    private final Action action;
+
+    AuthConfig(List<PolicyMatcher> policies, Action action) {
+      this.policies = Collections.unmodifiableList(policies);
+      this.action = action;
+    }
+  }
+
+  /** Instantiated with envoy policyMatcher configuration. */
+  public GrpcAuthorizationEngine(AuthConfig authConfig) {
+    this.authConfig = authConfig;
+  }
+
+  /** Return the auth decision for the request argument against the policies. */
+  public AuthDecision evaluate(Metadata metadata, ServerCall<?,?> serverCall) {
+    checkNotNull(metadata, "metadata");
+    checkNotNull(serverCall, "serverCall");
+    String firstMatch = null;
+    EvaluateArgs args = new EvaluateArgs(metadata, serverCall);
+    for (PolicyMatcher policyMatcher : authConfig.policies) {
+      if (policyMatcher.matches(args)) {
+        firstMatch = policyMatcher.name;
+        break;
+      }
+    }
+    Action decisionType = Action.DENY;
+    if (Action.DENY.equals(authConfig.action) == (firstMatch == null)) {
+      decisionType = Action.ALLOW;
+    }
+    log.log(Level.FINER, String.format("RBAC decision: {%s}, policy match: {%s}",
+        decisionType, firstMatch));
+    return AuthDecision.create(decisionType, firstMatch);
+  }
+
+  /**
+   * Implements a top level {@link Matcher} for a single RBAC policy configuration per envoy
+   * protocol:
+   * https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/rbac/v3/rbac.proto#config-rbac-v3-policy.
+   *
+   * <p>Currently we only support matching some of the request fields. Those unsupported fields are
+   * considered not match until we stop ignoring them.
+   */
+  static final class PolicyMatcher extends Matcher {
+    private final OrMatcher permissions;
+    private final OrMatcher principals;
+    private final String name;
+
+    /** Constructs a matcher for one RBAC policy. */
+    PolicyMatcher(String name, OrMatcher permissions, OrMatcher principals) {
+      this.name = name;
+      this.permissions = permissions;
+      this.principals = principals;
+    }
+
+    @Override
+    boolean matches(EvaluateArgs args) {
+      return permissions.matches(args) && principals.matches(args);
+    }
+  }
+
+  static final class AuthenticatedMatcher extends Matcher {
+    private final StringMatcher delegate;
+
+    AuthenticatedMatcher(StringMatcher delegate) {
+      this.delegate = delegate;
+    }
+
+    @Override
+    boolean matches(EvaluateArgs args) {
+      if (delegate == null) {
+        return true;
```

Comment:
Because i noticed this: https://github.com/grpc/grpc-java/blob/27b1641653b4c27232d473496a632db193735b74/xds/third_party/envoy/src/main/proto/envoy/config/rbac/v3/rbac.proto#L240, although there are other ways to do this, but then the parser needs to be careful to do the check.

Reply:
Uhh... this code matches _any_ connection, not just those that are authenticated.

In any case, that meaning of null should be documented in the constructor or similar.

Reply:
> this code matches any connection, not just those that are authenticated.

When the code reaches here in the auth interceptor, the protocol negotiation that does the authentication is already done. So, if we encounter an authenticated AuthConfig that is missing stringMatcher, it is a valid config and the auth engine consider it a match? This whole behaviour is strange to me, and seems inconsistent with other matcher config, like this is required:
https://github.com/grpc/grpc-java/blob/27b1641653b4c27232d473496a632db193735b74/xds/third_party/envoy/src/main/proto/envoy/type/matcher/path.proto#L25
Let me put this null to be allowed and put comment as well.

Reply:
Any authentication _if it is being done_ will already have happened. But this may be a plaintext connection.

Reply:
hm, to me plain connection also means it passed authentication, although we can also argue that certificate must be present. There might be a way to check the latter situation, but i may not want to go that way.

Reply:
No, no, no. There is no authentication at all with plaintext. Authentication means you have verified the remote party. Now, you could say all "necessary authentication" has occurred, but that is very different from saying the connection is authenticated.

I'm confused by "i may not want to go that way". How do our opinions change the meaning of the RBAC matchers?

Reply:
do we consider it authenticated or not if the connection is secured(`Grpc.TRANSPORT_ATTR_SSL_SESSION` is installed?) but the certificate content is unexpected to us (parsing exception, no certificate, etc.)